src/components/CheckableInput/CheckableInput.tsx

```diff
@@ -0,0 +1,52 @@
+import classNames from "classnames";
+import { TSFixMe } from "index";
+import { nanoid } from "nanoid";
+import PropTypes from "prop-types";
+import React, { useEffect, useRef, HTMLProps } from "react";
+
+type Props = {
+  type: string;
+  label: string;
+  required?: boolean;
+  indeterminate?: boolean;
+} & HTMLProps<HTMLElement>;
+
+const CheckableInput = ({
+  type,
+  label,
+  required = false,
+  indeterminate = false,
+  ...checkboxProps
+}: Props): JSX.Element => {
+  const inputId = useRef(nanoid());
+  // setting this type as any for now, otherwise it causes linting errors
+  const inputRef: TSFixMe = useRef(null);
+
+  useEffect(() => {
+    inputRef.current.indeterminate = indeterminate;
+  }, [indeterminate]);
+
+  return (
+    <label className={classNames(`p-${type}`, { "is-required": required })}>
+      <input
```

Comment:
How can I get the value of this input out? You might want to look at using a `forwardRef`.

How would I pass my own "input" element into here? Say something like using a `Formik` input? Without being able to do that it limits the functionality of this component.

Reply:
@hatched We were loosely looking at our current `Input` component when implementing this.

For the value, it just works as a standard input - you pass a `value`, `onChange` or whatever props the input needs to create a controlled input, something along these lines?

```js
const [checked, setChecked] = useState(false);

return <Checkbox checked={checked} onChange={() => setChecked(!checked)} />
```

It's a good call about `Formik`, we would need to have a look at how it works. Does our current `Input` component support it, or you implement your own components for that?

Reply:
@hatched Having a quick look at [Formik docs](https://formik.org/docs/overview#the-gist), it seems like they are passing props to inputs, which seems to be supported by our current `Checkbox` API, or am I missing something?

Do you have any example of what you mean by passing you own "input" element?

Reply:
Ah, just found that in next example:

<img width="727" alt="Screenshot 2021-04-16 at 07 09 50" src="https://user-images.githubusercontent.com/83575/114974505-cbf99280-9e82-11eb-8bf9-26ff9d0c8c22.png">


@hatched You have any suggestions what is the best way to pass such custom components into other components, via props, children?

This "looks nice" from JSX perspective, but is likely too open to any other child elements:
```js
<Checkbox>
   <Field type="checkbox" />
</Checkbox>
```

This seems to look a bit strange, not sure if it would even work or is a good practice in React:

```js
<Checkbox
   input={<Field type="checkbox" />}
/>
```

Reply:
This is how we use the react-component inputs with Formik in MAAS: https://github.com/canonical-web-and-design/maas-ui/blob/master/ui/src/app/base/components/FormikField/FormikField.tsx. At first glance it looks like the new component will continue work.

Reply:
Thanks @huwshimi, that's helpful.

Reply:
Thanks for the deep dive into this! I see now that there is `{...checkboxProps}` on line 38 that will remedy the original issue I had.